File: apps/api/src/modules/kpi/kpi.controller.ts
import {
  Controller,
  Get,
  Post,
  Patch,
  Delete,
  Body,
  Param,
  Query,
  ParseUUIDPipe,
  HttpCode,
  HttpStatus,
  UseGuards,
  UploadedFile,
  UseInterceptors,
} from '@nestjs/common';
import { FileInterceptor } from '@nestjs/platform-express';
import { ClerkJwtGuard } from '../auth/clerk-jwt.guard';
import { Session as Auth } from '../auth/decorators/session.decorator';
import type { TrpcAuth } from '@lotris/types';
import { KpiService } from './kpi.service';
import { KpiImportService } from './kpi-import.service';
import {
  CreateKpiDefinitionDto,
  UpdateKpiDefinitionDto,
  SetTeamTargetDto,
  CreateKpiAssignmentDto,
  CreateKpiAgreementDto,
  UpsertAgreementAreasDto,
  ImportColumnMappingDto,
  CreateActualDto,
} from './dto';

@Controller('api/v1/kpi')
@UseGuards(ClerkJwtGuard)
export class KpiController {
  constructor(
    private readonly kpi: KpiService,
    private readonly importSvc: KpiImportService,
  ) {}

  // ── KPI Definitions ────────────────────────────────────────────────────

  @Get('definitions')
  listDefinitions(@Auth() auth: TrpcAuth) {
    return this.kpi.listDefinitions(auth);
  }

  @Get('definitions/:id')
  getDefinition(@Auth() auth: TrpcAuth, @Param('id', ParseUUIDPipe) id: string) {
    return this.kpi.getDefinition(auth, id);
  }

  @Post('definitions')
  createDefinition(@Auth() auth: TrpcAuth, @Body() dto: CreateKpiDefinitionDto) {
    return this.kpi.createDefinition(auth, dto);
  }

  @Patch('definitions/:id')
  updateDefinition(
    @Auth() auth: TrpcAuth,
    @Param('id', ParseUUIDPipe) id: string,
    @Body() dto: UpdateKpiDefinitionDto,
  ) {
    return this.kpi.updateDefinition(auth, id, dto);
  }

  @Delete('definitions/:id')
  @HttpCode(HttpStatus.NO_CONTENT)
  archiveDefinition(@Auth() auth: TrpcAuth, @Param('id', ParseUUIDPipe) id: string) {
    return this.kpi.archiveDefinition(auth, id);
  }

  // ── Team Targets ───────────────────────────────────────────────────────

  @Get('definitions/:id/team-targets')
  getTeamTargets(@Auth() auth: TrpcAuth, @Param('id', ParseUUIDPipe) id: string) {
    return this.kpi.getTeamTargets(auth, id);
  }

  @Patch('definitions/:id/team-targets')
  setTeamTarget(
    @Auth() auth: TrpcAuth,
    @Param('id', ParseUUIDPipe) id: string,
    @Body() dto: SetTeamTargetDto,
  ) {
    return this.kpi.setTeamTarget(auth, id, dto);
  }

  // ── Engineer Assignments ───────────────────────────────────────────────

  @Get('assignments')
  listAssignments(
    @Auth() auth: TrpcAuth,
    @Query('engineerId') engineerId?: string,
    @Query('periodKey') periodKey?: string,
  ) {
    return this.kpi.listAssignments(auth, engineerId, periodKey);
  }

  @Post('assignments')
  createAssignment(@Auth() auth: TrpcAuth, @Body() dto: CreateKpiAssignmentDto) {
    return this.kpi.createAssignment(auth, dto);
  }

  // ── Agreements ─────────────────────────────────────────────────────────

  @Get('agreements')
  listAgreements(
    @Auth() auth: TrpcAuth,
    @Query('engineerId') engineerId?: string,
    @Query('periodKey') periodKey?: string,
  ) {
    return this.kpi.listAgreements(auth, engineerId, periodKey);
  }

  @Post('agreements')
  createAgreement(@Auth() auth: TrpcAuth, @Body() dto: CreateKpiAgreementDto) {
    return this.kpi.createAgreement(auth, dto);
  }

  @Get('agreements/:id')
  getAgreement(@Auth() auth: TrpcAuth, @Param('id', ParseUUIDPipe) id: string) {
    return this.kpi.getAgreementWithAreas(auth, id);
  }

  @Patch('agreements/:id/areas')
  upsertAreas(
    @Auth() auth: TrpcAuth,
    @Param('id', ParseUUIDPipe) id: string,
    @Body() dto: UpsertAgreementAreasDto,
  ) {
    return this.kpi.upsertAgreementAreas(auth, id, dto);
  }

  @Post('agreements/:id/submit')
  @HttpCode(HttpStatus.OK)
  submitAgreement(@Auth() auth: TrpcAuth, @Param('id', ParseUUIDPipe) id: string) {
    return this.kpi.submitAgreement(auth, id);
  }

  @Post('agreements/:id/accept')
  @HttpCode(HttpStatus.OK)
  acceptAgreement(@Auth() auth: TrpcAuth, @Param('id', ParseUUIDPipe) id: string) {
    return this.kpi.acceptAgreement(auth, id);
  }

  // ── Document Upload ────────────────────────────────────────────────────

  @Post('agreements/:id/upload')
  @UseInterceptors(FileInterceptor('file'))
  uploadAgreementFile(
    @Auth() auth: TrpcAuth,
    @Param('id', ParseUUIDPipe) id: string,
    @UploadedFile() file: Express.Multer.File,
  ) {
    return this.importSvc.parseUpload(auth, id, file);
  }

  @Post('agreements/:id/import')
  importAgreementRows(
    @Auth() auth: TrpcAuth,
    @Param('id', ParseUUIDPipe) id: string,
    @Body() dto: ImportColumnMappingDto,
  ) {
    return this.importSvc.importWithMapping(auth, id, dto);
  }

  // ── Scoring ────────────────────────────────────────────────────────────

  @Post('agreements/:id/score')
  @HttpCode(HttpStatus.OK)
  computeScore(@Auth() auth: TrpcAuth, @Param('id', ParseUUIDPipe) id: string) {
    return this.kpi.computeScore(auth, id);
  }

  @Get('agreements/:id/result')
  getResult(@Auth() auth: TrpcAuth, @Param('id', ParseUUIDPipe) id: string) {
    return this.kpi.getResult(auth, id);
  }

  // ── Actuals ────────────────────────────────────────────────────────────

  @Get('actuals')
  listActuals(
    @Auth() auth: TrpcAuth,
    @Query('engineerId') engineerId?: string,
    @Query('metricId') metricId?: string,
  ) {
    return this.kpi.listActuals(auth, engineerId, metricId);
  }

  @Post('actuals')
  createActual(@Auth() auth: TrpcAuth, @Body() dto: CreateActualDto) {
    return this.kpi.createActual(auth, dto);
  }
}

File: src/Lotris.Api/Controllers/KpiController.cs
using Lotris.Api.Auth;
using Lotris.Application.Kpi;
using Lotris.Contracts.Kpi;
using Microsoft.AspNetCore.Authorization;
using Microsoft.AspNetCore.Mvc;

namespace Lotris.Api.Controllers;

[ApiController]
[Authorize]
[Route("api/v1/kpi")]
public sealed class KpiController : ControllerBase
{
    private readonly KpiService _kpi;
    private readonly KpiImportService _import;

    public KpiController(KpiService kpi, KpiImportService import)
    {
        _kpi = kpi;
        _import = import;
    }

    [HttpGet("definitions")]
    public async Task<IActionResult> ListDefinitions(CancellationToken cancellationToken) =>
        Ok(await _kpi.ListDefinitionsAsync(HttpContext.GetLotrisSession(), cancellationToken));

    [HttpGet("definitions/{id:guid}")]
    public async Task<IActionResult> GetDefinition(Guid id, CancellationToken cancellationToken) =>
        Ok(await _kpi.GetDefinitionAsync(HttpContext.GetLotrisSession(), id, cancellationToken));

    [HttpPost("definitions")]
    public async Task<IActionResult> CreateDefinition(
        [FromBody] CreateKpiDefinitionRequest request,
        CancellationToken cancellationToken) =>
        Ok(await _kpi.CreateDefinitionAsync(HttpContext.GetLotrisSession(), request, cancellationToken));

    [HttpPatch("definitions/{id:guid}")]
    public async Task<IActionResult> UpdateDefinition(
        Guid id,
        [FromBody] UpdateKpiDefinitionRequest request,
        CancellationToken cancellationToken) =>
        Ok(await _kpi.UpdateDefinitionAsync(HttpContext.GetLotrisSession(), id, request, cancellationToken));

    [HttpDelete("definitions/{id:guid}")]
    [ProducesResponseType(StatusCodes.Status204NoContent)]
    public async Task<IActionResult> ArchiveDefinition(Guid id, CancellationToken cancellationToken)
    {
        await _kpi.ArchiveDefinitionAsync(HttpContext.GetLotrisSession(), id, cancellationToken);
        return NoContent();
    }

    [HttpGet("definitions/{id:guid}/team-targets")]
    public async Task<IActionResult> GetTeamTargets(Guid id, CancellationToken cancellationToken) =>
        Ok(await _kpi.GetTeamTargetsAsync(HttpContext.GetLotrisSession(), id, cancellationToken));

    [HttpPatch("definitions/{id:guid}/team-targets")]
    public async Task<IActionResult> SetTeamTarget(
        Guid id,
        [FromBody] SetTeamTargetRequest request,
        CancellationToken cancellationToken)
    {
        await _kpi.SetTeamTargetAsync(HttpContext.GetLotrisSession(), id, request, cancellationToken);
        return Ok(new { success = true });
    }

    [HttpGet("assignments")]
    public async Task<IActionResult> ListAssignments(
        [FromQuery] Guid? engineerId,
        [FromQuery] string? periodKey,
        CancellationToken cancellationToken) =>
        Ok(await _kpi.ListAssignmentsAsync(HttpContext.GetLotrisSession(), engineerId, periodKey, cancellationToken));

    [HttpPost("assignments")]
    public async Task<IActionResult> CreateAssignment(
        [FromBody] CreateKpiAssignmentRequest request,
        CancellationToken cancellationToken) =>
        Ok(await _kpi.CreateAssignmentAsync(HttpContext.GetLotrisSession(), request, cancellationToken));

    [HttpGet("agreements")]
    public async Task<IActionResult> ListAgreements(
        [FromQuery] Guid? engineerId,
        [FromQuery] string? periodKey,
        CancellationToken cancellationToken) =>
        Ok(await _kpi.ListAgreementsAsync(HttpContext.GetLotrisSession(), engineerId, periodKey, cancellationToken));

    [HttpPost("agreements")]
    public async Task<IActionResult> CreateAgreement(
        [FromBody] CreateKpiAgreementRequest request,
        CancellationToken cancellationToken) =>
        Ok(await _kpi.CreateAgreementAsync(HttpContext.GetLotrisSession(), request, cancellationToken));

    [HttpGet("agreements/{id:guid}")]
    public async Task<IActionResult> GetAgreement(Guid id, CancellationToken cancellationToken) =>
        Ok(await _kpi.GetAgreementWithAreasAsync(HttpContext.GetLotrisSession(), id, cancellationToken));

    [HttpPatch("agreements/{id:guid}/areas")]
    public async Task<IActionResult> UpsertAreas(
        Guid id,
        [FromBody] UpsertAgreementAreasRequest request,
        CancellationToken cancellationToken) =>
        Ok(await _kpi.UpsertAgreementAreasAsync(HttpContext.GetLotrisSession(), id, request, cancellationToken));

    [HttpPost("agreements/{id:guid}/submit")]
    public async Task<IActionResult> SubmitAgreement(Guid id, CancellationToken cancellationToken) =>
        Ok(await _kpi.SubmitAgreementAsync(HttpContext.GetLotrisSession(), id, cancellationToken));

    [HttpPost("agreements/{id:guid}/accept")]
    public async Task<IActionResult> AcceptAgreement(Guid id, CancellationToken cancellationToken) =>
        Ok(await _kpi.AcceptAgreementAsync(HttpContext.GetLotrisSession(), id, cancellationToken));

    [HttpPost("agreements/{id:guid}/upload")]
    [RequestSizeLimit(10 * 1024 * 1024)]
    public async Task<IActionResult> UploadAgreementFile(
        Guid id,
        IFormFile? file,
        CancellationToken cancellationToken)
    {
        if (file is null || file.Length == 0)
        {
            return BadRequest(new { message = "No file uploaded" });
        }

        await using var stream = file.OpenReadStream();
        return Ok(await _import.ParseUploadAsync(
            HttpContext.GetLotrisSession(),
            id,
            stream,
            file.FileName,
            cancellationToken));
    }

    [HttpPost("agreements/{id:guid}/import")]
    public async Task<IActionResult> ImportAgreementRows(
        Guid id,
        [FromBody] ImportColumnMappingRequest request,
        CancellationToken cancellationToken) =>
        Ok(await _import.ImportWithMappingAsync(HttpContext.GetLotrisSession(), id, request, cancellationToken));

    [HttpPost("agreements/{id:guid}/score")]
    public async Task<IActionResult> ComputeScore(Guid id, CancellationToken cancellationToken) =>
        Ok(await _kpi.ComputeScoreAsync(HttpContext.GetLotrisSession(), id, cancellationToken));

    [HttpGet("agreements/{id:guid}/result")]
    public async Task<IActionResult> GetResult(Guid id, CancellationToken cancellationToken) =>
        Ok(await _kpi.GetResultAsync(HttpContext.GetLotrisSession(), id, cancellationToken));

    [HttpGet("actuals")]
    public async Task<IActionResult> ListActuals(
        [FromQuery] Guid? engineerId,
        [FromQuery] Guid? metricId,
        CancellationToken cancellationToken) =>
        Ok(await _kpi.ListActualsAsync(HttpContext.GetLotrisSession(), engineerId, metricId, cancellationToken));

    [HttpPost("actuals")]
    public async Task<IActionResult> CreateActual(
        [FromBody] CreateActualRequest request,
        CancellationToken cancellationToken) =>
        Ok(await _kpi.CreateActualAsync(HttpContext.GetLotrisSession(), request, cancellationToken));
}
